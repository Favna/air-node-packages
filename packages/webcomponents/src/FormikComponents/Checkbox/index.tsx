import MUICheckbox, { CheckboxProps as MUICheckboxProps } from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import CheckBoxOutlineFilledIcon from '@material-ui/icons/CheckBox';
import CheckBoxOutlineBlankIcon from '@material-ui/icons/CheckBoxOutlineBlank';
import React, { ChangeEvent, forwardRef, ForwardRefExoticComponent } from 'react';
import css from './Checkbox.scss';

export interface CheckboxProps extends MUICheckboxProps {
  /** The name for the checkbox field */
  name: string;
  /** A tracker for the current value of the checkbox */
  value: string;
  /** Data-QA to send to the checkbox for testing */
  'data-qa'?: string;
  /**
   * The formik onChange event
   * @example `formikProps.handlechange`
   */
  onChange: (e: ChangeEvent<any>) => void;
  /**
   * The formik onBlur event
   * @example `formikProps.handleBlur`
   */
  onBlur: (...args: unknown[]) => void;
}

/**
 * Constructs a Checkbox with formik validation
 * @param props Props to pass to the Checkbox component
 * @example
 * ```jsx
 * <FieldArray
 *     name='fieldArrayNamne'
 *     render={arrayHelpers => (
 *     <Fragment>
 *         {arrayWithObjects.map((item, index) => (
 *         <Grid item xs={4} key={index}>
 *             <Tooltip title={`form.title.${item.name.toLowerCase()}`} placement='top'>
 *             <Checkbox
 *                 key={index}
 *                 value={item.name}
 *                 data-qa={`checkbox-${item.name.toLowerCase()}`}
 *                 name={`form.title.${item.name.toLowerCase()}`}
 *                 onChange={(e: ChangeEvent<{ checked: boolean }>) => {
 *                 if (e.target.checked) {
 *                     arrayHelpers.push(item.name);
 *                 } else {
 *                     const idx = formikProps.values.fieldName.indexOf(item.name);
 *                     arrayHelpers.remove(idx);
 *                 }
 *                 }}
 *                 onBlur={formikProps.handleBlur}
 *             />
 *             </Tooltip>
 *         </Grid>
 *         ))}
 *     </Fragment>
 *     )}
 * />
 * ```
 */
export const Checkbox: ForwardRefExoticComponent<CheckboxProps> = forwardRef(
  ({ name, value, onChange, onBlur, 'data-qa': dataQa, checked, ...props }, ref) => (
    <FormControlLabel
      ref={ref}
      control={
        <MUICheckbox
          {...props}
          icon={<CheckBoxOutlineBlankIcon className={css.checkBox} fontSize='small' />}
          checkedIcon={<CheckBoxOutlineFilledIcon fontSize='small' />}
          name={name}
          value={value}
          onChange={onChange}
          onBlur={onBlur}
          color='primary'
          data-qa={dataQa}
          checked={checked}
        />
      }
      label={name}
      classes={{ label: css.label }}
    />
  )
);
