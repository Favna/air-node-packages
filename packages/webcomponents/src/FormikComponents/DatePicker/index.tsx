import { KeyboardDatePicker } from '@material-ui/pickers';
import { fieldToKeyboardDatePicker, KeyboardDatePickerProps } from 'formik-material-ui-pickers';
import React, { FC } from 'react';
import css from './DatePicker.scss';

export interface DatePickerProps extends KeyboardDatePickerProps {
  /** The label shown in the datepicker */
  label: string;
  /** The minimum date to meet */
  minDate: Date;
}

/**
 * Constructs a Date Picker with formik validation
 * @param componentProps Props to pass to the date picker component
 * @example
 * ```jsx
 * <Field component={DatePicker} minDate={new Date('2012-01-01')} name='date' label='Sample Date' />
 * ```
 */
export const DatePicker: FC<DatePickerProps> = ({ minDate, label, ...props }) => {
  return (
    <KeyboardDatePicker
      {...props}
      {...fieldToKeyboardDatePicker(props)}
      animateYearScrolling
      autoOk
      format='DD/MM/YYYY'
      fullWidth
      InputAdornmentProps={{ position: 'start', className: css.datePickerIcon }}
      inputVariant='outlined'
      KeyboardButtonProps={{ 'aria-label': 'change date' }}
      label={label}
      minDate={minDate}
      type='date'
      variant='inline'
      InputProps={{
        classes: {
          root: css.input,
          notchedOutline: css.outline
        }
      }}
      InputLabelProps={{
        shrink: true,
        classes: { root: css.inputLabel }
      }}
      FormHelperTextProps={{
        // @ts-ignore
        component: 'div',
        classes: { root: css.errorLabel }
      }}
    />
  );
};
