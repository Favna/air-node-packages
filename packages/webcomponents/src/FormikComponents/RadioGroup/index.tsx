import { FormControlLabel, Radio } from '@material-ui/core';
import {
  RadioGroup as MUIRadioGroup,
  RadioGroupProps as MUIRadopGroupProps,
  useFieldToRadioGroup
} from 'formik-material-ui';
import React from 'react';
import css from './RadioGroup.scss';

/** The possible types for the option value */
export type OptionValue = string | number | boolean;

export interface Option<T extends OptionValue> {
  /** The value for the option */
  value: T;
  /** The label shown next to the radio button */
  label: string;
}

export interface RadioGroupProps<T extends OptionValue> extends MUIRadopGroupProps {
  /** The options for the radio buttons */
  options: Option<T>[];
  /** Whether this field is required */
  required?: boolean;
  /** Data-qa for the Radio Group */
  'data-qa'?: string;
}

/**
 * Constructs a radio button group with formik validation
 * @param componentProps Props to pass to the component
 * @example
 * ```jsx
 * <RadioGroup
 *   name='type'
 *   data-qa='sample-radio-group'
 *   required
 *   options={ [{ value: 'JOHN', label: 'John'}, { value: 'CONNOR', label: 'Connor'}] }
 * />
 * ```
 */
export const RadioGroup = <T extends OptionValue>({
  name,
  options,
  'data-qa': dataQa,
  ...props
}: RadioGroupProps<T>) => {
  const radioFieldProps = useFieldToRadioGroup({ name, ...props });

  return (
    <MUIRadioGroup {...props} {...radioFieldProps} name={name} row>
      {options.map((option, index) => (
        <FormControlLabel
          key={index}
          value={option.value}
          label={option.label}
          classes={{ label: css.label }}
          labelPlacement='end'
          control={<Radio data-qa={`${dataQa}-${option.value}`} required={props.required} color='primary' />}
        />
      ))}
    </MUIRadioGroup>
  );
};
