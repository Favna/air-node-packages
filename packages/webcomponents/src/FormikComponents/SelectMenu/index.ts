export { default as SelectMenu, SelectMenuProps, SelectMenuOption, SelectMenuOptionValues } from './SelectMenu';
