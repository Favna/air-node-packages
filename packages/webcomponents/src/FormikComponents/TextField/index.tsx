import { TextField as MUITextField } from '@material-ui/core';
import classnames from 'classnames';
import { useField } from 'formik';
import { fieldToTextField, TextFieldProps } from 'formik-material-ui';
import React, { FC } from 'react';
import css from './TextField.scss';

/**
 * Constructs a TextField with the formik validation
 * @param props Props to pass to the Textield component
 * @example
 * ```jsx
 * <Field
 *   component={TextField}
 *   name='name'
 *   type='text'
 *   label='Example'
 *   placeholder='Example placeholder'
 *   variant='outlined'
 *   data-qa='exampleDataQA'
 *   required
 * />
 * ```
 */
export const TextField: FC<TextFieldProps> = (props) => {
  const [{ value, onChange, onBlur }, { error, touched }] = useField(props.field.name);

  const textFieldHasErrors = Boolean(error) && touched;

  return (
    <MUITextField
      {...props}
      {...fieldToTextField(props)}
      value={value}
      label={props.label}
      placeholder={props.placeholder}
      onChange={onChange}
      onBlur={onBlur}
      margin='dense'
      variant='outlined'
      className={props.className}
      fullWidth
      autoFocus={props.autoFocus}
      required={props.required}
      InputProps={{
        classes: {
          root: css.input,
          notchedOutline: css.outline
        }
      }}
      classes={{ root: classnames([{ [css.errorInput]: textFieldHasErrors }]) }}
      InputLabelProps={{
        shrink: true,
        classes: { root: css.inputLabel }
      }}
      FormHelperTextProps={{
        // @ts-ignore
        component: 'div',
        classes: { error: css.errorLabel }
      }}
    />
  );
};
