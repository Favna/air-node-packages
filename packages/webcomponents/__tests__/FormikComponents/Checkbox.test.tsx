import MUICheckbox from '@material-ui/core/Checkbox/Checkbox';
import { shallow, ShallowWrapper } from 'enzyme';
import React, { createRef } from 'react';
import { Checkbox } from '../../src/FormikComponents/Checkbox';

const mockCheckboxOnChange = jest.fn();
const mockCheckboxOnBlur = jest.fn();

describe('Component Tests', () => {
  let checkBox: ShallowWrapper;

  beforeAll(() => {
    checkBox = shallow(
      <Checkbox name='testCheckbox' value='testCheckbox' onBlur={mockCheckboxOnBlur} onChange={mockCheckboxOnChange} />
    );
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('checkbox function called onChange', () => {
    const changeEvent = { event: { target: { value: true } } };
    checkBox
      .dive()
      .dive()
      .find(MUICheckbox)
      .simulate('change', changeEvent);
    expect(mockCheckboxOnChange).toHaveBeenCalledWith(changeEvent);
    expect(mockCheckboxOnChange).toHaveBeenCalledTimes(1);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const checkBox = shallow(
      <Checkbox name='testCheckbox' value='testCheckbox' onBlur={mockCheckboxOnBlur} onChange={mockCheckboxOnChange} />
    );
    expect(checkBox).toMatchSnapshot();
  });

  test('Passing custom checked property', () => {
    const checkBox = shallow(
      <Checkbox
        name='testCheckbox'
        value='testCheckbox'
        onBlur={mockCheckboxOnBlur}
        onChange={mockCheckboxOnChange}
        checked={['one', 'two', 'three'].includes('one')}
      />
    );
    expect(checkBox).toMatchSnapshot();
  });

  test('Passing a bunch of additional props', () => {
    const checkBox = shallow(
      <Checkbox
        autoCapitalize='yes'
        buttonRef={createRef()}
        checked={false}
        color='primary'
        disabled
        disableFocusRipple
        disableRipple
        disableTouchRipple
        name='testCheckbox'
        onAbort={jest.fn()}
        onBlur={mockCheckboxOnBlur}
        onChange={mockCheckboxOnChange}
        value='testCheckbox'
      />
    );
    expect(checkBox).toMatchSnapshot();
  });
});
