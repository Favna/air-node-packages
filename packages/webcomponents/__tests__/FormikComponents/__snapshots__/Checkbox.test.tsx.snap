// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Passing a bunch of additional props 1`] = `
<WithStyles(ForwardRef(FormControlLabel))
  classes={
    Object {
      "label": "label",
    }
  }
  control={
    <WithStyles(ForwardRef(Checkbox))
      autoCapitalize="yes"
      buttonRef={
        Object {
          "current": null,
        }
      }
      checked={false}
      checkedIcon={
        <Memo(CheckBoxIcon)
          fontSize="small"
        />
      }
      color="primary"
      disableFocusRipple={true}
      disableRipple={true}
      disableTouchRipple={true}
      disabled={true}
      icon={
        <Memo(CheckBoxOutlineBlankIcon)
          className="checkBox"
          fontSize="small"
        />
      }
      name="testCheckbox"
      onAbort={[MockFunction]}
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      value="testCheckbox"
    />
  }
  label="testCheckbox"
/>
`;

exports[`Snapshot Testing Passing custom checked property 1`] = `
<WithStyles(ForwardRef(FormControlLabel))
  classes={
    Object {
      "label": "label",
    }
  }
  control={
    <WithStyles(ForwardRef(Checkbox))
      checked={true}
      checkedIcon={
        <Memo(CheckBoxIcon)
          fontSize="small"
        />
      }
      color="primary"
      icon={
        <Memo(CheckBoxOutlineBlankIcon)
          className="checkBox"
          fontSize="small"
        />
      }
      name="testCheckbox"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      value="testCheckbox"
    />
  }
  label="testCheckbox"
/>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(FormControlLabel))
  classes={
    Object {
      "label": "label",
    }
  }
  control={
    <WithStyles(ForwardRef(Checkbox))
      checkedIcon={
        <Memo(CheckBoxIcon)
          fontSize="small"
        />
      }
      color="primary"
      icon={
        <Memo(CheckBoxOutlineBlankIcon)
          className="checkBox"
          fontSize="small"
        />
      }
      name="testCheckbox"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      value="testCheckbox"
    />
  }
  label="testCheckbox"
/>
`;
